storage/file_persister.go

```diff
@@ -43,3 +48,142 @@ func (l *LocalFilePersister) Persist(path string, data io.Reader) (err error) {
 }
+
+// RemoteFilePersister is to be used when files created by the browser module need
+// to be uploaded to a remote location. This uses a preSignedURLGetterURL to
+// retrieve one pre-signed URL. The pre-signed url is used to upload the file
+// to the remote location.
+type RemoteFilePersister struct {
+	preSignedURLGetterURL string
+	headers               map[string]string
+	basePath              string
+}
+
+// Persist will upload the contents of data to a remote location.
+func (r *RemoteFilePersister) Persist(ctx context.Context, path string, data io.Reader) (err error) {
+	pURL, err := r.getPreSignedURL(ctx, path)
+	if err != nil {
+		return fmt.Errorf("get presigned url: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, pURL, data)
+	if err != nil {
+		return fmt.Errorf("create upload request: %w", err)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("perform upload request: %w", err)
+	}
+	defer func() {
+		if _, cerr := io.Copy(io.Discard, resp.Body); cerr != nil && err == nil {
+			err = fmt.Errorf("drain upload response body: %w", cerr)
+		}
+
+		if cerr := resp.Body.Close(); cerr != nil && err == nil {
+			err = fmt.Errorf("close upload response body: %w", cerr)
+		}
+	}()
+
+	if err := checkStatusCode(resp); err != nil {
+		return fmt.Errorf("upload status code out of range: %w", err)
+	}
+
+	return err
+}
+
+func checkStatusCode(resp *http.Response) error {
+	if resp.StatusCode == 429 {
+		return fmt.Errorf("request is being rate limited")
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("status code is %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+// getPreSignedURL will retrieve the presigned url for the current file.
+func (r *RemoteFilePersister) getPreSignedURL(ctx context.Context, path string) (string, error) {
+	b, err := buildPresignedRequestBody(r.basePath, path)
+	if err != nil {
+		return "", fmt.Errorf("build request body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.preSignedURLGetterURL, b)
+	if err != nil {
+		return "", fmt.Errorf("create request: %w", err)
+	}
+
+	for k, v := range r.headers {
+		req.Header.Add(k, v)
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("perform request: %w", err)
+	}
+	defer func() {
+		if cerr := resp.Body.Close(); cerr != nil && err == nil {
+			err = fmt.Errorf("close response body: %w", cerr)
+		}
+	}()
+
+	if err := checkStatusCode(resp); err != nil {
+		return "", fmt.Errorf("status code out of range: %w", err)
+	}
+
+	return readResponseBody(resp)
+}
+
+func buildPresignedRequestBody(basePath, path string) (io.Reader, error) {
+	b := struct {
+		Service string `json:"service"`
+		Files   []struct {
+			Name string `json:"name"`
+		} `json:"files"`
+	}{
+		Service: "aws_s3",
+		Files: []struct {
+			Name string `json:"name"`
+		}{
+			{
+				Name: filepath.Join(basePath, path),
+			},
+		},
```

Comment:
It might make sense to define the `Files` as a type here to reduce repetition.

Reply:
They're not exactly the same since the request file object only has the `name` field, whereas the response `file` has a `name` and a `url`. But, if you have a strong option then how is best to implement that? Just `file` or both a `service` and a `file` which work for both request and response bodies? 

```go
type service struct {
	Service string `json:"service"`
	Files   []file `json:"files"`
	URLs    []file `json:"urls"`
}

type file struct {
	Name         string `json:"name"`
	PreSignedURL string `json:"pre_signed_url"` //nolint:tagliatelle
}
```

Reply:
No, it just caught my eye. It's fine if you think it's not the best way to use another type.

Reply:
Ok, i'm leaving it as it is since there's no real reason to move them out as their own types for now.